build.rs

```diff
@@ -0,0 +1,108 @@
+// following https://doc.rust-lang.org/cargo/reference/build-scripts.html
+use std::env;
+//use std::io::Write;         //needed for debugging
+//use std::path::PathBuf;     //needed for one approach
+//use std::fs;                //needed for one approach
+
+fn main() {
+    // This file (build.rs) needs to be in the package root.
+    // It is called before the linker and used to arrange for the linker to find the
+    // proper memory.x file for the MCU. The memory.x files are assumed to be in a
+    // directory memoryMaps/xxx/  where xxx is replaced by an one of
+
+    let mcus = [
+        "STM32F042",
+        "STM32F030XC",
+        "STM32F100",
+        "STM32F101",
+        "STM32F103",
+        "STM32F303XC",
+        "STM32F401",
+        "STM32F411",
+        "STM32F722",
+        "STM32H742",
+        "STM32L0X2",
+        "STM32L100",
+        "STM32L151",
+        "STM32L4X2",
+        "LM3S6965",
+        "GD32VF103CB",
+        "GD32VF103C8",
+        "GD32VF103_EXTRA",
+    ];
+
+    // For example,   memoryMaps/STM32F401/memory.x
+    // Note that the MCU string must be in upper case because it is also used to
+    // find the  CARGO_FEATURE_xxx environment variable (eg CARGO_FEATURE_STM32F401)
+    // which is in upper case.
+
+    // There are two possible appraches: one is to copy the appropriate memory.x file into the OUT_DIR
```

Comment:
the STM32 linker files are pretty static, the third option is to have one basic `memory.x` template that you fill in per processor.

this approach is fine however, though maybe this should be gated by the `examples` feature and the linker files could move from `memoryMaps/` to `examples/linker/`?

Reply:
I tried quite a few different things before I came up with this scheme. It really is just a directory of the static memory templates you suggest, organized in a way that's simple for build to find the right one.

I found schemes where the `memory.x` files were copied to somewhere, like root or the build target directory. Those were messy. They needed to be git ignored, and they caused errors if the wrong one was left in the wrong place. The target build directory is only by triple but the `memory.x` files are different for different MCUs. (e.g. `stm32f401` and `stm32f411` have different `memory.x` but the same target triple.) So parallel builds clobber one another if this is not done carefully.

I guess they could go in the `examples/` directory if they really are only for examples. Are there not cases where linking might be done elsewhere, for example in unit testing in `src/lib.rs`?

Reply:
> I guess they could go in the examples/ directory if they really are only for examples. Are there not cases where linking might be done elsewhere, for example in unit testing in src/lib.rs?

unit tests can't be run against `no_std` targets and my thinking is that the memory files aren't something _provided_ by this project to third parties

Reply:
OK, I'll move it to examples.